File: dotnet/Library/Okapi/Hashing.cs
using Okapi.Hashing.V1;

namespace Okapi.Hashing;

public static class Blake3
{
    public static Blake3HashResponse Hash(Blake3HashRequest request)
    {
        return Native.Call<Blake3HashRequest, Blake3HashResponse>(request, Native.blake3_hash);
    }

    public static Blake3KeyedHashResponse KeyedHash(Blake3KeyedHashRequest request)
    {
        return Native.Call<Blake3KeyedHashRequest, Blake3KeyedHashResponse>(request, Native.blake3_keyed_hash);
    }

    public static Blake3DeriveKeyResponse DeriveKey(Blake3DeriveKeyRequest request)
    {
        return Native.Call<Blake3DeriveKeyRequest, Blake3DeriveKeyResponse>(request, Native.blake3_derive_key);
    }
}

public static class Sha256
{
    public static SHA256HashResponse Hash(SHA256HashRequest request)
    {
        return Native.Call<SHA256HashRequest, SHA256HashResponse>(request, Native.sha256_hash);
    }
}

File: ruby/lib/hashing.rb
# frozen_string_literal: true

# Hashing module
module Hashing
  def self.blake3_hash(request)
    Okapi.verify_type(request, Hashing_V1::Blake3HashRequest)
    Okapi.ffi_call('blake3_hash', request, Hashing_V1::Blake3HashResponse)
  end

  def self.blake3_keyed_hash(request)
    Okapi.verify_type(request, Hashing_V1::Blake3KeyedHashRequest)
    Okapi.ffi_call('blake3_keyed_hash', request, Hashing_V1::Blake3KeyedHashResponse)
  end

  def self.blake3_derive_key(request)
    Okapi.verify_type(request, Hashing_V1::Blake3DeriveKeyRequest)
    Okapi.ffi_call('blake3_derive_key', request, Hashing_V1::Blake3DeriveKeyResponse)
  end

  def self.sha256_hash(request)
    Okapi.verify_type(request, Hashing_V1::SHA256HashRequest)
    Okapi.ffi_call('sha256_hash', request, Hashing_V1::SHA256HashResponse)
  end
end
